File: src/utils/file_handler.rb
require 'json'
require 'yaml'
require 'csv'
require 'fileutils'
require 'digest'

module BasicCli
  module Utils
    class FileHandler
      class FileError < StandardError; end
      class UnsupportedFormatError < FileError; end

      SUPPORTED_FORMATS = %w[.json .yaml .yml .csv .txt .log].freeze

      def self.read(filepath, format: nil)
        raise FileError, "File not found: #{filepath}" unless File.exist?(filepath)

        format ||= detect_format(filepath)
        content = File.read(filepath)

        case format
        when :json
          parse_json(content)
        when :yaml
          parse_yaml(content)
        when :csv
          parse_csv(content)
        else
          content
        end
      rescue StandardError => e
        raise FileError, "Failed to read #{filepath}: #{e.message}"
      end

      def self.write(filepath, data, format: nil, pretty: true)
        format ||= detect_format(filepath)

        FileUtils.mkdir_p(File.dirname(filepath))

        content = case format
                  when :json
                    pretty ? JSON.pretty_generate(data) : data.to_json
                  when :yaml
                    data.to_yaml
                  when :csv
                    generate_csv(data)
                  else
                    data.to_s
                  end

        File.write(filepath, content)
        true
      rescue StandardError => e
        raise FileError, "Failed to write #{filepath}: #{e.message}"
      end

      def self.copy(source, destination, create_dirs: true)
        raise FileError, "Source file not found: #{source}" unless File.exist?(source)

        FileUtils.mkdir_p(File.dirname(destination)) if create_dirs

        FileUtils.cp(source, destination)
        true
      rescue StandardError => e
        raise FileError, "Failed to copy file: #{e.message}"
      end

      def self.move(source, destination, create_dirs: true)
        raise FileError, "Source file not found: #{source}" unless File.exist?(source)

        FileUtils.mkdir_p(File.dirname(destination)) if create_dirs

        FileUtils.mv(source, destination)
        true
      rescue StandardError => e
        raise FileError, "Failed to move file: #{e.message}"
      end

      def self.delete(filepath)
        return false unless File.exist?(filepath)

        File.delete(filepath)
        true
      rescue StandardError => e
        raise FileError, "Failed to delete #{filepath}: #{e.message}"
      end

      def self.exists?(filepath)
        File.exist?(filepath)
      end

      def self.size(filepath)
        raise FileError, "File not found: #{filepath}" unless File.exist?(filepath)

        File.size(filepath)
      end

      def self.checksum(filepath, algorithm: :sha256)
        raise FileError, "File not found: #{filepath}" unless File.exist?(filepath)

        digest = case algorithm
                 when :md5
                   Digest::MD5
                 when :sha1
                   Digest::SHA1
                 when :sha256
                   Digest::SHA256
                 when :sha512
                   Digest::SHA512
                 else
                   raise ArgumentError, "Unsupported algorithm: #{algorithm}"
                 end

        digest.file(filepath).hexdigest
      end

      def self.stats(filepath)
        raise FileError, "File not found: #{filepath}" unless File.exist?(filepath)

        stat = File.stat(filepath)
        {
          size: stat.size,
          modified_at: stat.mtime,
          created_at: stat.ctime,
          accessed_at: stat.atime,
          permissions: stat.mode.to_s(8)[-3..],
          directory: stat.directory?,
          file: stat.file?,
          readable: stat.readable?,
          writable: stat.writable?,
          executable: stat.executable?
        }
      end

      def self.watch(filepath)
        raise FileError, "File not found: #{filepath}" unless File.exist?(filepath)

        last_mtime = File.mtime(filepath)

        loop do
          sleep 0.1
          current_mtime = File.mtime(filepath)

          if current_mtime > last_mtime
            yield(filepath, current_mtime)
            last_mtime = current_mtime
          end
        end
      end

      def self.atomic_write(filepath, data, format: nil)
        temp_file = "#{filepath}.tmp.#{Process.pid}"

        begin
          write(temp_file, data, format: format)
          FileUtils.mv(temp_file, filepath)
          true
        ensure
          FileUtils.rm_f(temp_file)
        end
      end

      def self.detect_format(filepath)
        ext = File.extname(filepath).downcase

        case ext
        when '.json'
          :json
        when '.yaml', '.yml'
          :yaml
        when '.csv'
          :csv
        else
          :text
        end
      end

      def self.parse_json(content)
        JSON.parse(content, symbolize_names: true)
      rescue JSON::ParserError => e
        raise FileError, "Invalid JSON: #{e.message}"
      end

      def self.parse_yaml(content)
        YAML.safe_load(content, permitted_classes: [Symbol, Date, Time], aliases: true)
      rescue Psych::SyntaxError => e
        raise FileError, "Invalid YAML: #{e.message}"
      end

      def self.parse_csv(content)
        CSV.parse(content, headers: true).map(&:to_h)
      rescue CSV::MalformedCSVError => e
        raise FileError, "Invalid CSV: #{e.message}"
      end

      def self.generate_csv(data)
        return '' if data.empty?

        if data.is_a?(Array) && data.first.is_a?(Hash)
          CSV.generate do |csv|
            csv << data.first.keys
            data.each { |row| csv << row.values }
          end
        else
          data.to_s
        end
      end
    end
  end
end

File: src/rust/utils/file_handler.rs
use anyhow::{bail, Context, Result};
use csv::{Reader, Writer};
use serde::{Deserialize, Serialize};
use sha1::Sha1;
use sha2::{Digest, Sha256, Sha512};
use std::collections::HashMap;
use std::fs::{self, File};
use std::io::{Read, Write as _};
use std::path::Path;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum FileError {
    #[error("File not found: {0}")]
    NotFound(String),
    
    #[error("Failed to read {0}: {1}")]
    ReadError(String, String),
    
    #[error("Failed to write {0}: {1}")]
    WriteError(String, String),
    
    #[error("Invalid JSON: {0}")]
    InvalidJson(String),
    
    #[error("Invalid YAML: {0}")]
    InvalidYaml(String),
    
    #[error("Invalid CSV: {0}")]
    InvalidCsv(String),
    
    #[error("Unsupported format: {0}")]
    UnsupportedFormat(String),
    
    #[error("File operation failed: {0}")]
    OperationFailed(String),
}

pub struct FileHandler;

impl FileHandler {
    pub fn read<P: AsRef<Path>>(path: P) -> Result<String> {
        let path = path.as_ref();
        if !path.exists() {
            bail!(FileError::NotFound(path.display().to_string()));
        }

        fs::read_to_string(path)
            .with_context(|| format!("Failed to read file: {:?}", path))
    }

    pub fn write<P: AsRef<Path>>(path: P, content: &str) -> Result<()> {
        let path = path.as_ref();
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }

        fs::write(path, content)
            .with_context(|| format!("Failed to write file: {:?}", path))
    }

    pub fn read_json<T, P>(path: P) -> Result<T>
    where
        T: for<'de> Deserialize<'de>,
        P: AsRef<Path>,
    {
        let content = Self::read(&path)?;
        serde_json::from_str(&content)
            .map_err(|e| FileError::InvalidJson(e.to_string()).into())
    }

    pub fn write_json<T, P>(path: P, data: &T, pretty: bool) -> Result<()>
    where
        T: Serialize,
        P: AsRef<Path>,
    {
        let json = if pretty {
            serde_json::to_string_pretty(data)?
        } else {
            serde_json::to_string(data)?
        };

        Self::write(path, &json)
    }

    pub fn read_yaml<T, P>(path: P) -> Result<T>
    where
        T: for<'de> Deserialize<'de>,
        P: AsRef<Path>,
    {
        let content = Self::read(&path)?;
        serde_yaml::from_str(&content)
            .map_err(|e| FileError::InvalidYaml(e.to_string()).into())
    }

    pub fn write_yaml<T, P>(path: P, data: &T) -> Result<()>
    where
        T: Serialize,
        P: AsRef<Path>,
    {
        let yaml = serde_yaml::to_string(data)?;
        Self::write(path, &yaml)
    }

    pub fn read_csv<P>(path: P) -> Result<Vec<HashMap<String, String>>>
    where
        P: AsRef<Path>,
    {
        let file = File::open(path.as_ref())?;
        let mut reader = Reader::from_reader(file);
        let headers = reader.headers()?.clone();
        
        let mut records = Vec::new();
        for result in reader.records() {
            let record = result?;
            let mut map = HashMap::new();
            for (i, field) in record.iter().enumerate() {
                if let Some(header) = headers.get(i) {
                    map.insert(header.to_string(), field.to_string());
                }
            }
            records.push(map);
        }
        
        Ok(records)
    }

    pub fn write_csv<P>(path: P, data: &[HashMap<String, String>]) -> Result<()>
    where
        P: AsRef<Path>,
    {
        if data.is_empty() {
            return Self::write(path, "");
        }

        let file = File::create(path.as_ref())?;
        let mut writer = Writer::from_writer(file);
        
        // Write headers
        let headers: Vec<_> = data[0].keys().cloned().collect();
        writer.write_record(&headers)?;
        
        // Write data
        for row in data {
            let record: Vec<_> = headers.iter().map(|h| row.get(h).unwrap_or(&String::new()).clone()).collect();
            writer.write_record(&record)?;
        }
        
        writer.flush()?;
        Ok(())
    }

    pub fn copy<P, Q>(source: P, destination: Q) -> Result<()>
    where
        P: AsRef<Path>,
        Q: AsRef<Path>,
    {
        let source = source.as_ref();
        let destination = destination.as_ref();

        if !source.exists() {
            bail!(FileError::NotFound(source.display().to_string()));
        }

        if let Some(parent) = destination.parent() {
            fs::create_dir_all(parent)?;
        }

        fs::copy(source, destination)?;
        Ok(())
    }

    pub fn move_file<P, Q>(source: P, destination: Q) -> Result<()>
    where
        P: AsRef<Path>,
        Q: AsRef<Path>,
    {
        let source = source.as_ref();
        let destination = destination.as_ref();

        if !source.exists() {
            bail!(FileError::NotFound(source.display().to_string()));
        }

        if let Some(parent) = destination.parent() {
            fs::create_dir_all(parent)?;
        }

        fs::rename(source, destination)?;
        Ok(())
    }

    pub fn delete<P: AsRef<Path>>(path: P) -> Result<bool> {
        let path = path.as_ref();
        if !path.exists() {
            return Ok(false);
        }

        fs::remove_file(path)?;
        Ok(true)
    }

    pub fn exists<P: AsRef<Path>>(path: P) -> bool {
        path.as_ref().exists()
    }

    pub fn size<P: AsRef<Path>>(path: P) -> Result<u64> {
        let path = path.as_ref();
        if !path.exists() {
            bail!(FileError::NotFound(path.display().to_string()));
        }

        let metadata = fs::metadata(path)?;
        Ok(metadata.len())
    }

    pub fn checksum<P: AsRef<Path>>(path: P, algorithm: &str) -> Result<String> {
        let path = path.as_ref();
        if !path.exists() {
            bail!(FileError::NotFound(path.display().to_string()));
        }

        let mut file = File::open(path)?;
        let mut buffer = Vec::new();
        file.read_to_end(&mut buffer)?;

        let hash = match algorithm {
            "md5" => {
                let digest = md5::compute(&buffer);
                format!("{:x}", digest)
            }
            "sha1" => {
                let mut hasher = Sha1::new();
                hasher.update(&buffer);
                format!("{:x}", hasher.finalize())
            }
            "sha256" => {
                let mut hasher = Sha256::new();
                hasher.update(&buffer);
                format!("{:x}", hasher.finalize())
            }
            "sha512" => {
                let mut hasher = Sha512::new();
                hasher.update(&buffer);
                format!("{:x}", hasher.finalize())
            }
            _ => bail!("Unsupported algorithm: {}", algorithm),
        };

        Ok(hash)
    }

    pub fn stats<P: AsRef<Path>>(path: P) -> Result<FileStats> {
        let path = path.as_ref();
        if !path.exists() {
            bail!(FileError::NotFound(path.display().to_string()));
        }

        let metadata = fs::metadata(path)?;
        
        Ok(FileStats {
            size: metadata.len(),
            modified_at: metadata.modified()?,
            created_at: metadata.created().ok(),
            accessed_at: metadata.accessed().ok(),
            is_directory: metadata.is_dir(),
            is_file: metadata.is_file(),
            #[cfg(unix)]
            permissions: {
                use std::os::unix::fs::PermissionsExt;
                format!("{:o}", metadata.permissions().mode() & 0o777)
            },
            #[cfg(not(unix))]
            permissions: String::from("N/A"),
        })
    }

    pub fn atomic_write<P: AsRef<Path>>(path: P, content: &str) -> Result<()> {
        let path = path.as_ref();
        let temp_path = path.with_extension(format!("tmp.{}", std::process::id()));

        Self::write(&temp_path, content)?;
        fs::rename(&temp_path, path)?;

        Ok(())
    }
}

#[derive(Debug)]
pub struct FileStats {
    pub size: u64,
    pub modified_at: std::time::SystemTime,
    pub created_at: Option<std::time::SystemTime>,
    pub accessed_at: Option<std::time::SystemTime>,
    pub is_directory: bool,
    pub is_file: bool,
    pub permissions: String,
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    #[test]
    fn test_read_write() {
        let dir = TempDir::new().unwrap();
        let file_path = dir.path().join("test.txt");

        FileHandler::write(&file_path, "Hello, World!").unwrap();
        let content = FileHandler::read(&file_path).unwrap();

        assert_eq!(content, "Hello, World!");
    }

    #[test]
    fn test_json_operations() {
        #[derive(Serialize, Deserialize, PartialEq, Debug)]
        struct TestData {
            name: String,
            value: i32,
        }

        let dir = TempDir::new().unwrap();
        let file_path = dir.path().join("test.json");

        let data = TestData {
            name: "Test".to_string(),
            value: 42,
        };

        FileHandler::write_json(&file_path, &data, true).unwrap();
        let loaded: TestData = FileHandler::read_json(&file_path).unwrap();

        assert_eq!(data, loaded);
    }

    #[test]
    fn test_copy_file() {
        let dir = TempDir::new().unwrap();
        let source = dir.path().join("source.txt");
        let dest = dir.path().join("dest.txt");

        FileHandler::write(&source, "test content").unwrap();
        FileHandler::copy(&source, &dest).unwrap();

        assert!(dest.exists());
        assert_eq!(FileHandler::read(&dest).unwrap(), "test content");
    }

    #[test]
    fn test_move_file() {
        let dir = TempDir::new().unwrap();
        let source = dir.path().join("source.txt");
        let dest = dir.path().join("dest.txt");

        FileHandler::write(&source, "move me").unwrap();
        FileHandler::move_file(&source, &dest).unwrap();

        assert!(!source.exists());
        assert!(dest.exists());
        assert_eq!(FileHandler::read(&dest).unwrap(), "move me");
    }

    #[test]
    fn test_delete_file() {
        let dir = TempDir::new().unwrap();
        let file_path = dir.path().join("delete_me.txt");

        FileHandler::write(&file_path, "temporary").unwrap();
        let result = FileHandler::delete(&file_path).unwrap();

        assert!(result);
        assert!(!file_path.exists());
    }

    #[test]
    fn test_checksum() {
        let dir = TempDir::new().unwrap();
        let file_path = dir.path().join("checksum.txt");

        FileHandler::write(&file_path, "Hello World").unwrap();

        let sha256 = FileHandler::checksum(&file_path, "sha256").unwrap();
        assert_eq!(sha256.len(), 64); // SHA256 is 64 hex chars

        let md5 = FileHandler::checksum(&file_path, "md5").unwrap();
        assert_eq!(md5.len(), 32); // MD5 is 32 hex chars
    }
}